File: src/ChpokkWeb/SystemFiles/Templates/ProjectTemplates/VisualBasic/Silverlight/1028/BusinessApplication/Helpers/DataFieldExtensions.vb
Imports System
Imports System.Runtime.CompilerServices
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Data

Public Module DataFieldExtensions
    ''' <summary>
    ''' 用另一個控制項取代 <see cref="DataField" />'s <see cref="TextBox" /> 控制項並且更新繫結。
    ''' </summary>
    ''' <param name="field">要取代其 <see cref="TextBox"/> 的 <see cref="DataField"/>。</param>
    ''' <param name="newControl">您要設定成 <see cref="DataField.Content" /> 的新控制項。</param>
    ''' <param name="dataBindingProperty">要使用於資料繫結之控制項的屬性。</param>        
    ''' <param name="bindingSetupFunction">
    ''' 選擇性 <c>Action</c>，可用來變更剛產生的繫結上的參數，然後再將它套用到 <paramref name="newControl"/>。
    ''' </param>
    <Extension()> _
    Public Sub ReplaceTextBox(ByVal field As DataField, ByVal newControl As FrameworkElement, ByVal dataBindingProperty As DependencyProperty, _
                              Optional ByVal bindingSetupFunction As Action(Of Binding) = Nothing)
        If field Is Nothing Then
            Throw New ArgumentNullException("field")
        End If

        If newControl Is Nothing Then
            Throw New ArgumentNullException("newControl")
        End If

        ' 複製現有的繫結來建構新繫結，並且傳遞給使用者指定的函式 (如果有)
        Dim newBinding As Binding = field.Content.GetBindingExpression(TextBox.TextProperty).ParentBinding.CreateCopy()

        If bindingSetupFunction IsNot Nothing Then
            bindingSetupFunction(newBinding)
        End If

        ' 取代欄位
        newControl.SetBinding(dataBindingProperty, newBinding)
        field.Content = newControl
    End Sub

    ''' <summary>
    ''' 從另一個 <see cref="Binding"/> 物件複製所有屬性來建立新 <see cref="Binding"/> 物件。
    ''' </summary>
    ''' <param name="binding">要複製屬性值的來源 <see cref="Binding"/></param>
    ''' <returns>新的 <see cref="Binding"/> 物件。</returns>
    <Extension()> _
    Public Function CreateCopy(ByVal binding As Binding) As binding
        If binding Is Nothing Then
            Throw New ArgumentNullException("binding")
        End If

        Dim newBinding As New Binding()

        newBinding.BindsDirectlyToSource = binding.BindsDirectlyToSource
        newBinding.Converter = binding.Converter
        newBinding.ConverterParameter = binding.ConverterParameter
        newBinding.ConverterCulture = binding.ConverterCulture
        newBinding.Mode = binding.Mode
        newBinding.NotifyOnValidationError = binding.NotifyOnValidationError
        newBinding.Path = binding.Path
        newBinding.UpdateSourceTrigger = binding.UpdateSourceTrigger
        newBinding.ValidatesOnExceptions = binding.ValidatesOnExceptions

        If binding.ElementName IsNot Nothing Then
            newBinding.ElementName = binding.ElementName
        ElseIf binding.RelativeSource IsNot Nothing Then
            newBinding.RelativeSource = binding.RelativeSource
        Else
            newBinding.Source = binding.Source
        End If

        Return newBinding
    End Function
End Module

File: src/ChpokkWeb/SystemFiles/Templates/ProjectTemplates/CSharp/Silverlight/1028/BusinessApplication/Helpers/DataFieldExtensions.cs
namespace $safeprojectname$
{
    using System;
    using System.Windows;
    using System.Windows.Controls;
    using System.Windows.Data;

    /// <summary>
    ///     提供在 <see cref="DataField"/> 上執行作業的擴充方法。
    /// </summary>
    public static class DataFieldExtensions
    {
        /// <summary>
        /// 用另一個控制項取代 <see cref="DataField" />'s <see cref="TextBox" /> 控制項並且更新繫結。
        /// </summary>
        /// <param name="field">要取代其 <see cref="TextBox"/> 的 <see cref="DataField"/>。</param>
        /// <param name="newControl">您要設定成 <see cref="DataField.Content" /> 的新控制項。</param>
        /// <param name="dataBindingProperty">要使用於資料繫結之控制項的屬性。</param>        
        public static void ReplaceTextBox(this DataField field, FrameworkElement newControl, DependencyProperty dataBindingProperty)
        {
            field.ReplaceTextBox(newControl, dataBindingProperty, null);
        }

        /// <summary>
        /// 用另一個控制項取代 <see cref="DataField" />'s <see cref="TextBox" /> 控制項並且更新繫結。
        /// </summary>
        /// <param name="field">要取代其 <see cref="TextBox"/> 的 <see cref="DataField"/>。</param>
        /// <param name="newControl">您要設定成 <see cref="DataField.Content" /> 的新控制項。</param>
        /// <param name="dataBindingProperty">要使用於資料繫結之控制項的屬性。</param>        
        /// <param name="bindingSetupFunction">
        ///  選擇性 <see cref="Action"/>，可用來變更剛產生的繫結上的參數，然後再將它套用到 <paramref name="newControl"/>
        /// </param>
        public static void ReplaceTextBox(this DataField field, FrameworkElement newControl, DependencyProperty dataBindingProperty, Action<Binding> bindingSetupFunction)
        {
            if (field == null)
            {
                throw new ArgumentNullException("field");
            }

            if (newControl == null)
            {
                throw new ArgumentNullException("newControl");
            }

            // 複製現有的繫結來建構新繫結，並且針對呼叫端要執行的任何變更將它傳遞給 bindingSetupFunction。
            Binding newBinding = field.Content.GetBindingExpression(TextBox.TextProperty).ParentBinding.CreateCopy();

            if (bindingSetupFunction != null)
            {
                bindingSetupFunction(newBinding);
            }

            // 取代欄位
            newControl.SetBinding(dataBindingProperty, newBinding);
            field.Content = newControl;
        }

        /// <summary>
        /// 從另一個 <see cref="Binding"/> 物件複製所有屬性來建立新 <see cref="Binding"/> 物件。
        /// </summary>
        /// <param name="binding">要複製屬性值的來源 <see cref="Binding"/></param>
        /// <returns>新的 <see cref="Binding"/> 物件。</returns>
        private static Binding CreateCopy(this Binding binding)
        {
            if (binding == null)
            {
                throw new ArgumentNullException("binding");
            }

            Binding newBinding = new Binding()
            {
                BindsDirectlyToSource = binding.BindsDirectlyToSource,
                Converter = binding.Converter,
                ConverterParameter = binding.ConverterParameter,
                ConverterCulture = binding.ConverterCulture,
                Mode = binding.Mode,
                NotifyOnValidationError = binding.NotifyOnValidationError,
                Path = binding.Path,
                UpdateSourceTrigger = binding.UpdateSourceTrigger,
                ValidatesOnExceptions = binding.ValidatesOnExceptions
            };

            if (binding.ElementName != null)
            {
                newBinding.ElementName = binding.ElementName;
            }
            else if (binding.RelativeSource != null)
            {
                newBinding.RelativeSource = binding.RelativeSource;
            }
            else
            {
                newBinding.Source = binding.Source;
            }

            return newBinding;
        }
    }
}
